pyvec/naucse.python.cz
# API versioning + session serials

Here's a draft of what API versioning can look like, plus a relatively easy change that introduces a new API version, 0.1.

## Session.serial

Currently we number sessions from 1 (Lekce 1, Lekce 2, ...), but I've had some demand (not only from my course) to number from 0:
![image](https://user-images.githubusercontent.com/302922/53299418-071b5080-383a-11e9-8060-d1b6957e7dc9.png)

That session number is not actually number: like page or chapter "numbers" in a book, I can imagine some sessions to use `i`, `ii`, `iii`, `iv`, or even nothing at all. So, I call it `serial` rather than “serial number”.

In the new API 0.1, `serial` is expected on input (i.e. what naucse_render gives us). If a session doesn't have a serial, the “Lekce 1” won't appear on it.
For the old API, 0.0 (which all courses now use), serial is auto-generated, as before.

With this change, the code can dump to either API 0.0 or 0.1, i.e. with or without `serial`. But the web only exports API 0.1 (under the old `v0/` URLs). That should not be a problem, it's just an added field.

## API versioning in converters

Load and dump operations pass a `context`, which includes the API version loaded/dumped.
A `VersionField` is added, which can select a field based on the version. So, when defining a new field, you can [specify the version it was added in](https://github.com/pyvec/naucse.python.cz/commit/52a5b4e29f53d4f70e8adf7f82f082d915c97d77#diff-0dc2efafae5d3f3f3787bd3950738463R479) and the new behavior. When updating the behavior, leave the old Field and add a new one with a new version.

The tests are updated with version-specific expected YAML output, so we can test that output for old API versions has not changed.

## How to test

- Install a version of `naucse_render` that outputs API 0.1 (https://github.com/pyvec/naucse_render/pull/8): 
  ```
   pipenv install git+https://github.com/encukou/naucse_render@serial#egg=naucse_render
   ```
- Observe that courses no longer say “Lekce 1” in the session list
- Add `- serial: 0` to one of the course YAMLs, and observe that sessions are numbered starting with “Lekce 0”.
